### ISSUE TITLE
Language System Tags for Chinese (Traditional)

### ISSUE THREAD

**Issue by hfhchan** [MicrosoftDocs/typography-issues#654]
There are currently three language system tags being registered which are related to the use of the Chinese Traditional script:

ZHT Chinese Traditional
ZHH Chinese, Hong Kong SAR
ZHTM Chinese, Macao SAR

These are confusing, because 'ZHT' naming is script based, while 'ZHH' / 'ZHTM' is locale based, and somehow implies Taiwan is the default locale for Chinese Traditional.  That's not the case.

Taiwan and Hong Kong have different set of standardized orthographies.  For example, in Source Han Sans 1.x series, only characters with the Taiwan standardized forms were shipped, which caused considerable uproar for a font targeted for pan-CJK use. In SHS 2.000, 1,750 glyphs were removed from 1.004 were made to make space for new glyphs, most of which were to include Hong Kong specific forms.

On the other hand, for most fonts produced by local foundries targeting Taiwan, Hong Kong and Macao market, often they do not adhere to either the Taiwanese standardized orthography or Hong Kong standardized orthography specifically.  They will often either follow the (1) Traditional orthography (傳統字形、舊字形), which are the forms which have been in use since the appearance of the Ming/Song style since the Ming dynasty, also often termed "Kangxi orthography", which is also shared by Japan for characters outside the Jōyō Kanji list, and Korea; (2) "Modern" orthography (新字形), where the forms are adjusted to follow handwriting more closely; (3) a mixture of both, where in the same character common radical components are using the traditional orthography, while other components are using the modern orthography.

A considerable number of users in Taiwan are also rejecting the common orthography approach and strong proponents of using the standardized orthography in Taiwan.  There should be a way for fonts to deliver the correct form to users' preference based on language system tags without overwriting Hong Kong user's preferences.

Lastly, when Chinese (Simplified) is used in Hong Kong and Macao, it's expected that the characters will be rendered in the orthography used by Chinese Simplified, so ZHH and ZHTM are not expected to be invoked for language tags zh-Hans-HK and zh-Hans-MO.

I suggest:
(1) An additional ZHTT to be registered for Chinese Traditional, Taiwan
(2) An additional ZHTH to be registered for Chinese Traditional, Hong Kong SAR, and ZHH to be deprecated in favour of ZHTH
(3) ZHTM be renamed to Chinese Traditional, Macao SAR

This has the following implications:
(1) Fonts using an orthography that's suitable for Traditional Chinese areas should continue to use the ZHT language system tag.
(2) Fonts strictly adhering to the Taiwanese standardized orthography should add support for ZHTT language system tag in addition to the current ZHT tag, or only for the ZHTT tag when fallback is widely available.
(3) Implementations should add fallback to activate the ZHT tag when ZHTH, ZHTM or ZHTT tags are not available in the chosen font.

**Comment by hfhchan** [comment]
Here's a codepen testing browser behaviour:
https://codepen.io/hfhchan/full/abZprGo

Right now on Mac OS X, Firefox Nightly (84.0a1) and Google Chrome (86.0) exhibit the following behaviour:
zh-Hant: ZHT activated
zh-Hant-TW: ZHT activated
zh-TW: ZHT activated
zh-Hant-HK: ZHH activated
zh-HK: ZHH activated

Expected behaviour:
zh-Hant: ZHT activated
zh-Hant-TW: ZHTT or ZHT activated
zh-TW: ZHTT or ZHT activated
zh-Hant-HK: ZHH or ZHT activated
zh-HK: ZHH or ZHT activated

**Comment by hfhchan** [comment]
Here's a codepen testing browser behaviour:
https://codepen.io/hfhchan/full/abZprGo

Right now on Mac OS X, Google Chrome (86.0) exhibit the following behaviour:
zh-Hant: ZHT activated
zh-Hant-TW: ZHT activated
zh-TW: ZHT activated
zh-Hant-HK: ZHH activated
zh-HK: ZHH activated

Mac OS X, Firefox Nightly (84.0a1):
zh-Hant: ZHT activated
zh-Hant-TW: ZHT activated
zh-TW: ZHT activated
zh-Hant-HK: neither ZHT / ZHH activated
zh-HK: neither ZHT / ZHH activated

This de-facto equivalents zh-Hant to zh-TW.

Expected new behaviour:
zh-Hant: ZHT activated
zh-Hant-TW: ZHTT or ZHT activated
zh-TW: ZHTT or ZHT activated
zh-Hant-HK: ZHTH or ZHH or ZHT activated
zh-HK: ZHTH or ZHH or ZHT activated

**Comment by hfhchan** [comment]
Google Chrome for Android (86.0), Firefox for Android (81.1.5) also exhibit same behaviour for Google Chrome (86.0) on Mac OS X:
zh-Hant: ZHT activated
zh-Hant-TW: ZHT activated
zh-TW: ZHT activated
zh-Hant-HK: ZHH activated
zh-HK: ZHH activated

This is on a phone with English (Hong Kong) and Traditional Chinese (Hong Kong) in the language list, the bare minimum is zh-Hant should fallback to ZHH first instead of ZHT.

**Comment by PeterCon** [comment]
@kenlunde I'm interested in your thoughts on this.

**Comment by PeterCon** [comment]
>I suggest:
>(1) An additional ZHTT to be registered for Chinese Traditional, Taiwan
>(2) An additional ZHTH to be registered for Chinese Traditional, Hong Kong SAR, and ZHH to be deprecated in favour of ZHTH
>(3) ZHTM be renamed to Chinese Traditional, Macao SAR

(3) seems straightforward.

As for (1) and (2)...

>zh-Hant: ZHT activated
>zh-Hant-TW: ZHT activated
>zh-TW: ZHT activated

That seems to suggest that ZHT is already being treated as meaning "Chinese Traditional, Taiwan", so no need to add a new ZHTT tag.

>Lastly, when Chinese (Simplified) is used in Hong Kong and Macao, it's expected that the characters will be rendered in the orthography used by Chinese Simplified, so ZHH and ZHTM are not expected to be invoked for language tags zh-Hans-HK and zh-Hans-MO.

So, therefore, ZHH doesn't need to mean anything other than "Chinese Traditional, Hong Kong SAR"

Thus, for (1) and (2), wouldn't it make sense to clarify the existing ZHH and ZHT tags as follows?

ZHT = Chinese Traditional, Taiwan
ZHH = Chinese Traditiona, Hong Kong SAR

**Comment by dscorbett** [comment]
It would not be a good idea to narrow the scope of 'ZHT ' from zh-Hant to zh-TW because there are other uses of traditional Chinese beyond Taiwan, Hong Kong, and Macao. For example, HarfBuzz maps lzh (Literary Chinese) to 'ZHT '.

**Comment by PeterCon** [comment]
So far, for OT 1.8.4, have added "Traditional" to the descriptions for ZHH and ZHTM.

<https://docs.microsoft.com/en-us/typography/opentype/otspec184beta/delta/languagetags.html>

For ZHT (or ZHTT), waiting on further discussion.

**Comment by PeterCon** [comment]
>It would not be a good idea to narrow the scope of 'ZHT ' from zh-Hant to zh-TW... For example, HarfBuzz maps lzh (Literary Chinese) to 'ZHT '.

If there is a need to distinguish typography for modern Hant usage in Taiwan from usage in other regions, it's worth considering what is most expedient. 

Of course, there is a ton of data language tagged using one of zh-TW, zh-Hant or zh-Hant-TW. But there are also a lot of existing fonts designed for use in Taiwan, and if any existing fonts make use of language system tags to differentiate typographic conventions for Taiwan versus anything else, then ZHT is going to be the language system tag that's used. A new tag ZHTT could be registered, but that doesn't help existing fonts in the wild that already use ZHT; and it doesn't help existing software that maps language tags zh-TW/zh-Hant/zh-Hant-TW to the language system tag ZHT.

As for distinguishing typography for Classical / "Literary" Chinese, while there may be existing software implementations that map lzh to ZHT, 
* I would have to imaging that mapping was done assuming there isn't a need (or available way) to distinguish typographic conventions for Classical Chinese from those for modern usage in Taiwan.
* I think it's safe to assume there is far, far less content language tagged with lzh than there is content language tagged zh-TW/zh-Hant/zh-Hant-TW.
* I would guess that there are far fewer fonts designed to support specifically Classical Chinese (i.e., with forms that are distinct from any modern Hant usage) than there are fonts designed to support modern usage in Taiwan.

In other words, if there's a need in fonts to distinguish forms needed for modern Taiwan usage from forms needed for other regions *and also forms required for Classical Chinese*, then it would probably be *far less disruptive* to register a new language system tag for Classical Chinese forms than to register a new language system tag ZHTT for modern Taiwan.

**Comment by hfhchan** [comment]
> A new tag ZHTT could be registered, but that doesn't help existing fonts in the wild that already use ZHT; and it doesn't help existing software that maps language tags zh-TW/zh-Hant/zh-Hant-TW to the language system tag ZHT.

It also doesn't break compatibility, because the premise of introducing ZHTT is for supporting fallback for ZHTT/ZHTH/ZHTM to ZHT, so existing implementations would effectively be in "always fallback" mode.

Note most commercial fonts (i.e. hundreds by MonoType, Dynacomware, Arphic, etc) designed for the Taiwanese market don't ship with the Taiwan standard forms. Microsoft Jhenghei, PingFang TC and Source Han Sans/Source Han Serif are the exceptions.

Suppose Classical / "Literary" Chinese is registered with ZHTL, having ZHT/ZHH/ZHTM fall back to ZHTL sounds non-obvious.

**Comment by hfhchan** [comment]
> I would guess that there are far fewer fonts designed to support specifically Classical Chinese (i.e., with forms that are distinct from any modern Hant usage) than there are fonts designed to support modern usage in Taiwan.

Typically, Literary Chinese is not particularly distinguished from Standard Chinese in terms of typographic conventions; traditional or (non-standard) modern forms are used by most large publishers.

I don't think registering a Literary Chinese language system tag is appropriate at all for representing the traditional orthography at all. For the bulk of history when Literary Chinese was actually used, the Song script didn't even exist yet.

**Comment by hfhchan** [comment]
> That seems to suggest that ZHT is already being treated as meaning "Chinese Traditional, Taiwan", so no need to add a new ZHTT tag.

I'm arguing that's the wrong approach.  Vendors shouldn't be falling back zh-Hant as if it were zh-TW, (btw neither should they be treating zh as zh-Hans).  When the locale is missing from the author specified language tag, it should default to user's language preferences, instead of making a best guess based on population size.

**Comment by dscorbett** [comment]
I think y’all are focusing too much on my off-hand example of Classical Chinese. My point was merely that there are other uses of zh-Hant; another example is that traditional characters are in use in emigrant populations. I think that 'ZHT ' should remain zh-Hant in general. A font could use 'ZHT ' for zh-Hant-TW or for zh-Hant of any other region, as the font designer chooses. If there is a need for a tag for zh-Hant-TW specifically distinguished from zh-Hant (which I have no opinion on) then it would make sense to introduce a ZHTT tag.

**Comment by PeterCon** [comment]
>My point was merely that there are other uses of zh-Hant;

I'm not entirely sure why we'd be focusing on IETF language tags here. Yes, content is language tagged, and based on the language tagging a layout implementation can select an OT language system tag for selection of alternate feature tables in OT Layout processing. But the main question here is what OT language system tags need to be registered, not how content is language tagged or how implementations map from an IETF language tag to an OTL language system tag.

I have no opinion on whether there is a need in OT Layout processing to select distinct feature tables for Traditional-Chinese-in-Taiwan typography from any other Traditional Chinese typography. The proposal to register ZHTT, to me, implies an assumption that there is such a need, but I'm not qualified to evaluate that need. I would only question, if there is such a need, whether the best solution is a new ZHTT tag for "Chinese, Traditional, Taiwan", or rather if it would be better to interpret ZHT as being *that* tag and register one or more tags for whatever forms of Traditional Chinese typography (besides ZHH and ZHTM) need to be distinguished from it.

**Comment by dscorbett** [comment]
I write “zh-Hant” because it shorter than “Chinese written in traditional characters” and less ambiguous than “traditional Chinese”. This is not meant to imply that OpenType language system tags should specifically focus on IETF language tags.

**Comment by PeterCon** [comment]
@hfhchan You referred several times to "fallback". I'd like to get clarification on what you mean.

The term "fallback" is generally used in relation to *resource selection*. In this context, people commonly talk about *font fallback*: if an otherwise-selected font is not capable of displaying a string, an alternate "fallback" font is selected that is capable of displaying the string. If that's the kind of fallback you have in mind, then OT language system tags are not intended for that purpose. Rather, data in the 'meta' table should be used for font selection.

OT Layout language system tags are used in selection of feature tables. Now, feature tables are types of resources, and potentially the concept of fallback can apply here. E.g., based on some criteria (such as content language tags) a layout engine would prefer to select feature tables associated with OT language system tag ABCD, but if ABCD is not available in a given font, then other feature tables would used as a fallback. A GSUB or GPOS table necessarily has a default language system, hence default feature tables. And in most implementations, I would expect a simple fallback algorithm: if ABCD isn't supported in a font, fallback to the default language system. 

But potentially a layout engine might use a more-involved language system fallback: If ABCD language system isn't supported in the given font, then search for EFGH language system; else use the default.

In your various mentions of "fallback" it's not clear what you mean.

>This is on a phone with English (Hong Kong) and Traditional Chinese (Hong Kong) in the language list, the bare minimum is zh-Hant should fallback to ZHH first instead of ZHT.

In this case, that doesn't sound to me like *fallback* at all: If the content is language tagged as zh-Hant, and the user context indicates Hong Kong, then the *preferred* (not fallback) OT language system (and associated feature tables) to select would be the ZHH language system.

>Browser vendors shouldn't be falling back zh-Hant as if it were equivalent to zh-TW... When the locale is missing from the author specified language tag, and guessing the locale based on content is not available/feasible, it should at least fallback according to user's language preferences, instead of hard coding a fallback path based on population size.

This seems similar, except in this case you haven't called out what type of resource selection is involved: *Which font?* *Which OT language system?*

>the premise of introducing ZHTT is for supporting fallback for ZHTT/ZHTH/ZHTM to ZHT, so existing implementations would effectively be in "always fallback" mode.

>Suppose Classical / "Literary" Chinese is registered with ZHTL, having ZHT/ZHH/ZHTM fall back to ZHTL sounds non-obvious.

Here, your using language system tags, so maybe you do mean fallback for OTL language system selection during OT layout processing—the second type of fallback I described above. But it's not actually clear to me if that's really what you have in mind.


So, clarification would be helpful. Thanks.

**Comment by hfhchan** [comment]
Thank you for bearing with me for my inaccurate use of terminology.

> But potentially a layout engine might use a more-involved language system fallback: If ABCD language system isn't supported in the given font, then search for EFGH language system; else use the default.

Yes, this is what I'm referring to.

> In this case, that doesn't sound to me like fallback at all: If the content is language tagged as zh-Hant, and the user context indicates Hong Kong, then the preferred (not fallback) OT language system (and associated feature tables) to select would be the ZHH language system.

Yes.  But I also think that if ZHH locl cannot be found in a certain font, but ZHT can, then ZHT should also be selected instead.

> This seems similar, except in this case you haven't called out what type of resource selection is involved: Which font? Which OT language system?

I'm not sure if browsers are canonicalizing zh-Hant to zh-Hant-TW/zh-TW first, and then asking for ZHT to be activated, or zh-Hant directly activates ZHT.  Either way, both are not ideal behaviours.

> Here, your using language system tags, so maybe you do mean fallback for OTL language system selection during OT layout processing—the second type of fallback I described above. But it's not actually clear to me if that's really what you have in mind.

Yes, OTL language system selection during OT layout processing.

**Comment by hfhchan** [comment]
Let me step back a bit and also explain the use-cases.

The printed Chinese (Traditional) script has undergone a few changes in orthography.

First, the printed form, or Song or Ming script, becomes widespread in the Ming dynasty.  The characters used resemble the handwriting form, or Regular script, but have notable differences.  These printed forms are gradually refined over the centuries, departing more and more from the Regular script.  A notable milestone is when the Kangxi Dictionary is published, which is deemed to be the authoritative style for the Qing dynasty.  Hence the traditional printed forms are also known as "Kangxi forms".  These forms are the basis for all the lead type produced by Japan and Hong Kong, used throughout the whole East Asia in the 20th century.  These forms are also still used in Korea and for non Joyo-kanji in Japan.

After the WWII, both Taiwan and Hong Kong carried out character reforms which brought the orthography of glyphs closer to the handwriting standards, but in different ways.  Commerical font vendors also did their own shapes to make printed forms closer to handwriting forms.

Most commercially produced fonts are one either in the (1) Traditional printed form (aka "old form"/"traditional form"/"print form" 舊字形/傳統字形/印刷字形); (2) Mixed form where common radicals are in the traditional printed form and some other components are modified to have a form slightly closer to handwriting; (3) Components are drawn in a form similar to handwriting (aka "modern form"/"regular-script-ized form" 新字形/楷化字形); (4) Similar to (3) but characters are drawn adhering to either Taiwan govt's or Hong Kong govt's specific standard.

Only a small set of fonts are in category (4), nearly all lead type and wood block type belong in (1), and most newly produced commercial fonts are in set (3).

Nearly all traffic signs and a lot of commercial advertising / printed matter in Hong Kong is in (1).  A lot of old traffic signs and commercial advertising / printed matter in Taiwan is also in (1).

Currently, as Source Han Sans has implemented, the Taiwan govt's standard forms (4) and Hong Kong govt's standard forms (4) are activated by ZHT and ZHH respectively.  However, (4) is only one of forms used in Taiwan and Hong Kong.  In fact, before Microsoft Jhenghei and PingFang were introduced, fonts following (1) and (3) were used by default on Windows and Mac OS X.  Many other phone vendors also shipped with fonts in (1)/(2) shape until Android 9.

The biggest required use-case is for supporting (1) for Taiwan, Hong Kong and Macau users.  This is the orthography most Traditional Chinese users have grown up with.

One way to do it is to narrow the Taiwan govt's standardized orthography and Hong Kong govt's standardized orthography to separate language systems, and keep the existing ZHT tag as "any orthography (most likely (1)), suitable for Taiwan & HK & Macau".  That also means that there should be some way for users to force zh-TW tagged content and zh-HK tagged content to activate ZHT only and not ZHTH/ZHTT.

Another way is doing this via another mechanism like jp78/jp83/jp90/jp04, but how they interact with locl I am not familiar with.

**Comment by PeterCon** [comment]
>Thank you for bearing with me...

Thanks for your clarifications on "fallback". It's helpful to understand that you do have in mind layout engines checking a font for a specific language system tag and, if not present, then falling back to a more general language system tag.

Note, though, that such fallback logic lies outside the OpenType spec and OTL registry: Language system tags shouldn't be registered because certain fallback logic is used, but because there are different conventions for typography practice for the same characters/scripts to be distinguished so that a single font can support the distinct conventions.

>Let me step back a bit...
That background is also helpful. Thanks.


>Commercial fonts produced in style (1) - (3) are produced to be pan-TC. They aren't customized specifically for Taiwan, Hong Kong or Macao.

And,

>For (3) and (4)... many font vendors only emphasize (4) as being adhering to government standards, but fonts in the (3) style are the actual ones that commercially sell.

Based on these comments, I get the following impressions:

* Having language system tags to distinguish forms for 4a or 4b versus 3 is not a particular need
* Nonetheless, ZHH and ZHTM seem to provide ways to distinguish 4 for Hong Kong and Macao.
* ZHT could, potentially, be used for 3—and IIUC, that's what you'd like—*except that Source Han Sans is triggering 4a by the ZHT language system*.

So, what forms does Source Han Sans present if the default language system is used?

But the above appears to be of secondary importance based on this:
>The biggest required use-case is for supporting (1) shape for Taiwan, Hong Kong and Macau users.

So, aside from what to do with regard to ZHT and distinguishing forms for 4a vs. 3, it seems like the _first_ concern should be how to support forms for 1 (Kangxi style).

Of course, Kangxi-specific fonts could be produced, but I gather you'd have in mind how to support 1 in pan-TC fonts. 

You could certainly register a language system tag for 1 (say, ZHKX). A question to consider, though, is how that would get activated. Apps potentially _could_ provide a way for users to select language systems directly, as they might features, but I don't know of any apps that support that, and at this point I'm not sure it would be the best thing to pursue. 

So, let's assume that language system tags are most likely to be selected on the basis of IETF language tagging of content. That means that, for a language system tag like ZHKX to be most useful, there would need to be a BCP 47 tag that could determine it: perhaps using a variant subtag, e.g., zh-Hant-kangxi. (Of course, lzh could by default map to ZHKX, but I'm assuming something used for modern content like road signs, etc., not Literary Chinese.) But are users likely to language-tag content using a special BCP 47 subtag?

>triggering traditional printed style glyphs via another mechanism like jp78/jp83/jp90/jp04

For 1, registering a _feature_ may be the most useful—for sake of discussion, let's call it zhkx: 
* The zhkx feature would need to be exposed in applications, but that's much more likely to happen than apps exposing language systems to direct user control.
* It doesn't require any special practice wrt language-tagging of content.

As for this concern,
> but how they interact with locl, I am not familiar with

A zhkx feature could override whatever effects are produced by use of the 'locl' feature and language system tags. So, for example, if the ZHH language system is selected and the glyph substituted by 'locl' for some character is a 4b HK-specific form, a substitution lookup triggered by the 'zhkx' feature could be process after the 'locl' substitution and replace the HK-specific glyph with a traditional glyph.


Coming back to the 3 vs. 4a distinction and the status of ZHT, it sounds like more needs to get clarified there, and that perhaps there's a question as to whether Source Han Sans is doing the right thing for ZHT, (presenting 4b forms rather then 3 forms)?

**Comment by hfhchan** [comment]
I need slightly more time to process this tonight, but if u remember correctly, Arial Unicode MS delivers 3 forms when locl with ZHT is activated.

**Comment by hfhchan** [comment]
I need slightly more time to process this tonight, but if I am not mistaken, Arial Unicode MS delivers 3 forms when locl with ZHT is activated.

**Comment by PeterConstable** [comment]
@hfhchan It's been a year. Are you still wanting to pursue something here, or can this be closed?

**Comment by hfhchan** [comment]
Yes. I'll get back ASAP.

**Comment by hfhchan** [comment]
Yes this issue is still valid, though I'm not sure how to proceed.

Also, China has just recently published the national standard GB/Z 40637-2021, which contains standard forms to be used for Chinese characters when reprinting historical books to be applied by publishers.  The forms stipulated differ from the Traditional Chinese forms as so-called Taiwanese and HK conventions.

CC: @eisoch

One of the main pain points that still exist is how we could map different preferences of orthographies to different language codes.

**Comment by hfhchan** [comment]
> Note, though, that such fallback logic lies outside the OpenType spec and OTL registry: Language system tags shouldn't be registered because certain fallback logic is used, but because there are different conventions for typography practice for the same characters/scripts to be distinguished so that a single font can support the distinct conventions.

Right.

> Having language system tags to distinguish forms for 4a or 4b versus 3 is not a particular need

I would not be inclined to group orthography 3 with orthography 4a / 4b.

Here's an analysis of fonts published by Dynacomware (華康):
- 華康細明體 which was originally in orthography 1, was shipped with Windows up until XP as 細明體 & 新細明體, and the commercially available version is now using orthography 2.
- 華康儷宋體 developed in 1989, developed for the Hong Kong market is using orthography 3.
- 華康標宋體 & 華康香港標準宋體 developed for Education purposes in orthography 4a and 4b separately.

This illustrates that there is a commercial need to distinguish the different orthographies. Whether or not that directly translates into a commercial need to pack them all in the same font is another issue, but at least there are clear cut lines for different typographic conventions.

Existing evidence suggests that there is a pressing need to distinguish 1 & 2 versus 4a/4b, as well as 3 versus 4b.

The need to distinguish 1 & 2 versus 4a/4b has wide support based on continual feedback in the Source Han Sans GitHub issues list, and the proliferation of projects such as the [Genyo series](https://github.com/ButTaiwan/genyo-font) of fonts.

The need to distinguish 3 versus 4b has support based on the existence of the [Chiron Sans HK](https://chiron-fonts.github.io/sans/) font project.

> Nonetheless, ZHH and ZHTM seem to provide ways to distinguish 4 for Hong Kong and Macao.

Yes.

> ZHT could, potentially, be used for 3—and IIUC, that's what you'd like—except that Source Han Sans is triggering 4a by the ZHT language system.

I think that's fine for Source Han Sans to serve 4a under ZHT based on their scope.  From a backwards compat and practical perspective if fonts only intent to support 4a and 4b, serving 4a under ZHT would be the reasonable choice as well.

Note: Arial Unicode MS already uses ZHT to trigger orthography 3, and Arial Unicode MS is by far more widely available than Source Han Sans.

> So, what forms does Source Han Sans present if the default language system is used?

Source Han Sans currently uses the J forms by default, but say if the 65K barrier was not an issue, I think they should use orthography 2 by default on non-CJK locales. But in that case they would not need to use ZHT (assuming ZHT means generic Chinese (Traditional)) tag at all.

> You could certainly register a language system tag for 1 (say, ZHKX). A question to consider, though, is how that would get activated. Apps potentially could provide a way for users to select language systems directly, as they might features, but I don't know of any apps that support that, and at this point I'm not sure it would be the best thing to pursue.

Yes this is the main pain point.

> So, let's assume that language system tags are most likely to be selected on the basis of IETF language tagging of content. That means that, for a language system tag like ZHKX to be most useful, there would need to be a BCP 47 tag that could determine it: perhaps using a variant subtag, e.g., zh-Hant-kangxi. (Of course, lzh could by default map to ZHKX, but I'm assuming something used for modern content like road signs, etc., not Literary Chinese.)

Unfortunately the vast amount of content produced and intended to be consumed using orthography 1/2 would be in Standard Written Chinese.  Also with the introduction of GB/Z 40637-2021, let's call that orthography 4c, that will predominantly apply to content tagged Literary Chinese. So variant subtags that apply to any Sinitic language (`zh`, `zh-Hant`, `zh-Hans`, `lzh`, `yue`, `cmn`, `nan`, etc, and probably also `ja` and to a lesser extent `ko`) would make sense.

> But are users likely to language-tag content using a special BCP 47 subtag?

I think authors are not likely to language tag content using a special BCP 47 subtag, because publishers would simply pick a different font, and web content authors likely would not understand this complexity.  Instead, the think the main use case we can solve here is how to cater to users who have a strong preference about which orthography should render in application UIs.

Supporting a special BCP 47 subtag is probably the least intrusive way to do it, and probably the most convenient.

A user on Android 9 and above can choose the locales "Arabic (United Arab Emirates)" ("ar-AE", identical to "ar-AE-u-nu-arab")  or "Arabic (United Arab Emirates - Western Numerals)" ("ar-AE-nu-latn").  Users who prefer the traditional orthography will likely be expecting a similar interaction. (On older versions of Java, "ar-AE" seems to be identical to "ar-AE-nu-latn" instead)

Plumbing already exists for subtags to apply to application UIs.

**Comment by NightFurySL2001** [comment]
Is there any updates to this? Given the recent publication of GB/Z 40637-2021, it is in more urgent need to separate ZHT from being a locale tag to a script tag. Fonts that currentlly implemented ZHT to zh-hant-TW can keep the mapping as default in addition to mapping it to ZHTT as the designer chose. It would be more beneficial to add/modify the tags as such:

| Locale | OT Tag (current) | OT Tag (proposed) |
|-|-|-|
| zh-hant (generic) | Default to ZHT | ZHT |
| zh-hant-CN | - | ZHTC/ZHTG? |
| zh-hant-TW | ZHT | ZHTT |
| zh-hant-HK | ZHK | ZHTH |
| zh-hant-MO | ZHTM | ZHTM |

In par with this, Chinese (Simplified) can use an additional tag too, which is ZHSC (which is redundant, but it can separate ZHS as being a script and ZHSC as being a locale).

Note: CNS11643 from Taiwan actually did defined some forms for Simplified Chinese characters which has significant differences than those in use in China, and those forms are included in system fonts such as Microsoft JhengHei. A locale tag ZHST will be appropriate in that case (but is not required right now).

| Locale | OT Tag (current) | OT Tag (proposed) |
|-|-|-|
| zh-hans (generic) | Default to ZHS | ZHS |
| zh-hans-CN | ZHS | ZHSC/ZHSG? |
| zh-hans-TW | - | ZHST |

**Comment by hfhchan** [comment]
Additional information: GB/Z 40637-2021 (古籍印刷通用字规范字形表) specifies 14250 required glyphs for printing glyphs used in historical books. The applicable area of the standard is include commercial reproduction of books published before 1911, as well as any modern publication using traditional Chinese characters.  It is substantially different from any existing Chinese (Simplified) or Chinese (Traditional) standard. It also specifies the Unicode Scalar Values that have to be used with those glyphs.

**Comment by NightFurySL2001** [comment]
Given the `name` table has changed to using Microsoft LCID at https://github.com/MicrosoftDocs/typography-issues/issues/1104, updating the OpenType language tags for SC and TC will be immensely helpful here. Also might be useful to add zh-MY for Malaysia under `Y` tag (`M` is taken by Macao/MO).

| Locale | OT Tag (current) | OT Tag (proposed) |
|-|-|-|
| zh-hant (generic) | Default to ZHT | ZHT |
| zh-hant-CN | - | ZHTC/ZHTG? |
| zh-hant-TW | ZHT | ZHTT |
| zh-hant-HK | ZHH | ZHTH |
| zh-hant-MO | ZHTM | ZHTM |
| zh-hant-MY | - | ZHTY |
|-|-|-|
| zh-hans (generic) | Default to ZHS | ZHS |
| zh-hans-CN | ZHS | ZHSC/ZHSG? |
| zh-hans-SG | - | ZHSS |
| zh-hans-MY | - | ZHSY |
| zh-hans-TW | - | ZHST |
